src/modules/mc_pos_control/mc_pos_control_main.cpp

```diff
@@ -417,80 +414,18 @@ MulticopterPositionControl::poll_subscriptions()
 {
-	if (_vehicle_land_detected.alt_max < 0.0f || !_home_pos.valid_alt || !_local_pos.v_z_valid) {
-		// there is no altitude limitation present or the required information not available
-		return;
-	}
-
-	// maximum altitude == minimal z-value (NED)
-	const float min_z = _home_pos.z + (-_vehicle_land_detected.alt_max);
-
-	if (_states.position(2) < min_z) {
-		// above maximum altitude, only allow downwards flight == positive vz-setpoints (NED)
-		setpoint.z = min_z;
-		setpoint.vz = math::max(setpoint.vz, 0.f);
-	}
-}
-
-void
-MulticopterPositionControl::set_vehicle_states(const float &vel_sp_z)
-{
-	if (_local_pos.timestamp == 0) {
-		return;
-	}
-
-	// only set position states if valid and finite
-	if (PX4_ISFINITE(_local_pos.x) && PX4_ISFINITE(_local_pos.y) && _local_pos.xy_valid) {
-		_states.position(0) = _local_pos.x;
-		_states.position(1) = _local_pos.y;
-
-	} else {
-		_states.position(0) = _states.position(1) = NAN;
-	}
-
-	if (PX4_ISFINITE(_local_pos.z) && _local_pos.z_valid) {
-		_states.position(2) = _local_pos.z;
-
-	} else {
-		_states.position(2) = NAN;
-	}
-
-	if (PX4_ISFINITE(_local_pos.vx) && PX4_ISFINITE(_local_pos.vy) && _local_pos.v_xy_valid) {
-		_states.velocity(0) = _local_pos.vx;
-		_states.velocity(1) = _local_pos.vy;
-		_states.acceleration(0) = _vel_x_deriv.update(_states.velocity(0));
-		_states.acceleration(1) = _vel_y_deriv.update(_states.velocity(1));
-
-	} else {
-		_states.velocity(0) = _states.velocity(1) = NAN;
-		_states.acceleration(0) = _states.acceleration(1) = NAN;
-		// reset derivatives to prevent acceleration spikes when regaining velocity
-		_vel_x_deriv.reset();
-		_vel_y_deriv.reset();
-	}
-
-	if (PX4_ISFINITE(_local_pos.vz) && _local_pos.v_z_valid) {
-		_states.velocity(2) = _local_pos.vz;
-
-		if (PX4_ISFINITE(vel_sp_z) && fabsf(vel_sp_z) > FLT_EPSILON && PX4_ISFINITE(_local_pos.z_deriv)) {
-			// A change in velocity is demanded. Set velocity to the derivative of position
-			// because it has less bias but blend it in across the landing speed range
-			float weighting = fminf(fabsf(vel_sp_z) / _param_mpc_land_speed.get(), 1.0f);
-			_states.velocity(2) = _local_pos.z_deriv * weighting + _local_pos.vz * (1.0f - weighting);
+	if (local_position.z_valid && (_vehicle_land_detected.alt_max > 0) && _home_pos.valid_alt) {
+		// maximum altitude == minimal z-value (NED)
+		const float min_z = _home_pos.z + (-_vehicle_land_detected.alt_max);
+
+		if (local_position.z < min_z) {
+			// above maximum altitude, only allow downwards flight == positive vz-setpoints (NED)
+			setpoint.z = min_z;
+			setpoint.vz = math::max(setpoint.vz, 0.f);
 		}
-
-		_states.acceleration(2) = _vel_z_deriv.update(_states.velocity(2));
-
-	} else {
-		_states.velocity(2) = _states.acceleration(2) = NAN;
-		// reset derivative to prevent acceleration spikes when regaining velocity
-		_vel_z_deriv.reset();
-	}
-
-	if (PX4_ISFINITE(_local_pos.heading)) {
-		_states.yaw = _local_pos.heading;
```

Comment:
on previous behavior there is update of yaw only if _local_pos.heading is finite, but now heading can be use without checking that.
Is that ok?

Reply:
It shouldn't be up to the position controller to independently try and validate every single field, most of the other consumers of this data don't perform the same checks, so if this ever did catch a problem you'd still be screwed.

Reply:
the question, is it going to bump an issue that might been hidden until now?
maybe it worth to force NAN on simulator and see what happen